File: controllers/IDP_controller/IDP_controller.cpp
#include <webots/Robot.hpp>
#include <webots/Motor.hpp>
#include <webots/LightSensor.hpp>
#include <webots/DistanceSensor.hpp>
#include <webots/GPS.hpp>
#include <webots/Compass.hpp>
#include <math.h>

#define TIME_STEP 32
#define MAX_SPEED 10

// All the webots classes are defined in the "webots" namespace
using namespace webots;

//Creating new robot instance
Robot *robot = new Robot();

//Get devices
Motor *leftMotor = robot->getMotor("Wheel_L");
Motor *rightMotor = robot->getMotor("Wheel_R");
Motor *leftArmMotor = robot->getMotor("Arm_L");
Motor *rightArmMotor = robot->getMotor("Arm_R");
GPS *gps = robot->getGPS("gps");
Compass *compass = robot->getCompass("compass");

//Initialsing US and IR sensors
DistanceSensor *ds[4];
char dsNames[4][20] = {"us_right","us_left","ir_left","ir_right"};
  
/*LightSensor *light_sensor = robot->getLightSensor("TEPT4400");
light_sensor->enable(TIME_STEP);
*/

//Initiate vector to store sensor readings
std::vector<std::vector<double>> dsValueScan(1000,std::vector<double> (4,0));

void rotate(double theta){

// set the target position of the motors
  leftMotor->setPosition(theta);
  rightMotor->setPosition(-theta);
  
  // set up the motor speeds at 50% of the MAX_SPEED.
   leftMotor->setVelocity(0.5 * MAX_SPEED);
   rightMotor->setVelocity(0.5 * MAX_SPEED);
}

void move_forwards() {

   leftMotor->setPosition(INFINITY);
   rightMotor->setPosition(INFINITY);

   leftMotor->setVelocity(0.5 * MAX_SPEED);
   rightMotor->setVelocity(0.5 * MAX_SPEED);
}

void move_backwards() {
   // set the target position of the motors
   leftMotor->setPosition(INFINITY);
   rightMotor->setPosition(INFINITY);

   // set up the motor speeds at 50% of the MAX_SPEED.
   leftMotor->setVelocity(-0.5 * MAX_SPEED);
   rightMotor->setVelocity(-0.5 * MAX_SPEED);
}

void move_position(double x){
   // set the target position of the motors
   leftMotor->setPosition(x);
   rightMotor->setPosition(x);
}

void rotate_CW() {
   // set the target position of the motors
   leftMotor->setPosition(INFINITY);
   rightMotor->setPosition(INFINITY);

   // set up the motor speeds at 50% of the MAX_SPEED.
   leftMotor->setVelocity(0.5 * MAX_SPEED);
   rightMotor->setVelocity(-0.5 * MAX_SPEED);
}

void rotate_ACW() {
   // set the target position of the motors
   leftMotor->setPosition(INFINITY);
   rightMotor->setPosition(INFINITY);

   // set up the motor speeds at 50% of the MAX_SPEED.
   leftMotor->setVelocity(-0.5*MAX_SPEED);
   rightMotor->setVelocity(0.5*MAX_SPEED);
}

void open_arms() {
     // set the target position of the motors
     leftArmMotor->setPosition(0.5);
     rightArmMotor->setPosition(-0.5);
}

void close_arms() {
   // set the target position of the motors
   leftMotor->setPosition(0.0);
   rightMotor->setPosition(0.0);
}

std::vector<double> getdsValues(){
    std::vector<double> dsValues(4);
    for (int i = 0; i < 4; i++){
      dsValues[i] = ds[i]->getValue();
    }
    return dsValues;
}

double get_bearing_in_degrees(const double *north) {
    double rad = atan2(north[0], north[2]);
    double bearing = (rad - 1.5708) / M_PI * 180.0;
    if (bearing < 0.0){
      bearing = bearing + 360.0;
    }
    return bearing;
}

/*void rotate_theta(const double *north, double theta) { //function to be used in scan_on_spot
  
  double initial_bearing = get_bearing_in_degrees(north);
  double angle_rotated = 0.0;
  
  while (angle_rotated < theta) {
    rotate_CW();
    angle_rotated++;
    double bearing = get_bearing_in_degrees(north);
    std::cout << bearing;
    
    if ((bearing - initial_bearing) >= 0.0) {
      angle_rotated = bearing - initial_bearing;
    }
    if ((bearing - initial_bearing) < 0.0) {
      angle_rotated = bearing + 360.0 - initial_bearing;
    }
    
  } 
}
*/

std::vector<double> findBlocks(std::vector<double> &dsValueScan);
   
int main(int argc, char **argv) {
    
    //Enabling navigation
    gps->enable(TIME_STEP);
    compass->enable(TIME_STEP);
  
    //Retrieving device tags and enabling with refresh time step
    for(int i = 0; i < 4; i++){
      ds[i] = robot->getDistanceSensor(dsNames[i]);
      ds[i]->enable(TIME_STEP); //TIME_STEP defines rate at which sensor is refreshed
    }
    
   int j = 0;
   
   while (robot->step(TIME_STEP) != -1){
   
      /*double initial_bearing = get_bearing_in_degrees(north);
      rotate(18.2);
      double final_bearing = get_bearing_in_degrees(north);*/
      
      std::vector<double> Values = getdsValues();
      j++;
      rotate_ACW();
      
      for(int i = 0; i<4; i++){
        std::cout << Values[i]<< ",";
       }
       
      std::cout << "\n";
      
      if(j==100){
        break;
       }
     
      /*std::cout << initial_bearing << "," << final_bearing << "\n";
      
      double initial_position[3];
      if (i == 0){
      initial_position[0] = gps->getValues()[0];
      initial_position[1] = gps->getValues()[1];
      initial_position[2] = gps->getValues()[2];
     const double initial_north = compass->getValues();
    };
      
      double bearing = get_bearing_in_degrees(north);
      //std::cout << bearing << std::endl;
      //std::cout <<  position[0] << ", "<< position[1] << ", "<< position[2] << std::endl;
      //std::cout <<  north[0] << ", "<< north[1] << ", "<< north[2] << std::endl;
      //std::cout <<  initial_position[0] << ", "<< initial_position[1] << ", "<< initial_position[2] << std::endl;
      //std::cout << i << std::endl;
      //i++;
      */
   }
      
   delete robot;
   return 0;
 }

File: controllers/IDP_controller/IDP_controller.py
from controller import Robot, Motor
import numpy as np

TIME_STEP = 64
MAX_SPEED = 10

class Mybot():
		def __init__(self):
        # create a robot
        robot = Robot()

        # get devices
        self.us_right = robot.getDevice("us_right")
        self.us_left = robot.getDevice("us_left")
        self.ir = robot.getDevice("ir_right")
        self.lightSensor = robot.getDevice("TEPT4400")
        self.motor_left = robot.getDevice("Wheel_L")
        self.motor_right = robot.getDevice("Wheel_R")
        self.arm_left = robot.getDevice("Arm_L")
        self.arm_right = robot.getDevice("Arm_R")
        self.compass = robot.getCompass("compass")
    
    def move(self, speed):
        self.motor_left.setPosition(float('inf'))
        self.motor_right.setPosition(float('inf'))
        self.motor_left.setVelocity(speed * MAX_SPEED)
        self.motor_right.setVelocity(speed * MAX_SPEED)
    
    def open_arms(self):
        self.arm_left.setPosition(1)
        self.arm_right.setPosition(-1)
          
    #Rowan time
    def close_arms(self):
        self.arm_left.setPosition(0) #those arms do be looking kinda closed though
        self.arm_right.setPosition(0)   
    
    #Rowan time
    def getSensorValues(self):
				sensorValues = [[0,0],[0,0]]
        for i in range (2):
        		sensorValues[i] = ds[i].getValue()/5700 ####NOT sure on this line
        
        lookUpTable = ds[2].getLookupTable() ##Or this one
        lookUpTableSize = ds[2].getLookupTableSize() ###Or this one. WHat is ds?
    		
        distanceActual = 0
        
        
        
      #are you kidding me this deleted everything I was sat for 45 minutes doing -Liza
      #Yea so I was working on getting sensor lookup tables and values, cool, gonna go to supo and then get back to whatever state this is in
    
    # add other robot movement

    # sensor and bearing functions
    # 
    # In your code in cpp you passed the get bearing in degrees function
    # as an argument since it was declared later and you couldn't use it
    # here but in python this is unnecessary

    def getRawSensorValues(self):
        ir = self.ir.getValue()
        us_r_value = self.us_right.getValue()
        us_l_value = self.us_left.getValue()
        sensor_values = [ir,us_r_value,us_l_value]
        return sensor_values
    
    def 

    def getBearingInDegrees(self):
        north = compass.getValues()						
        rad = np.arctan(north[0]/north[2])
        bearing = rad/np.pi*180.0
        if bearing<0:
						bearing += 360
        return bearing
    
    #Rowan wrote this function
    #Function to rotate to a given bearing
    #Queries: I'm rusty on classes- is self.motor_left etc correct or is it just motor_left? I think I have done it right
    def rotateTheta(self, theta, inital_bearing):
				self.motor_left.setVelocity(0.5*MAX_SPEED)                                 #Set velocity of each motor to 0.5*max speed or -0.5*max speed
				self.motor_right.setVelocity(-0.5*MAX_SPEED) 
    		while True:                                                                #While true loop
        		bearing = getBearingInDegrees()                                        #Get bearing of block from where I am 
        		if ((bearing - initial_bearing) >= 0:                                  #If i'm not pointing at block, angle to rotated is different
								angle_rotated = bearing - inital bearing
            else: 
            		angle_rotated = bearing - inital bearing + 360
            if angle_rotated > theta:                                              #Woo the angle has been found
            		self.motor_left.setVelocity(0)
                self.motor_right.setVelocity(0)
                break
     
    #Rowan Time
    #Not sure the difference between this and rotate theta. Asked cindy though ;)
    #Again, self.motor unsure
		def rotateUntilBearing(self, target_bearing, initial_bearing):
    		angle_rotated = 0
        if target_bearing > initial_bearing:
        		self.left_motor.setVelocity(0.1*MAX_SPEED)
            self.right_motor.setVelocity(-0.1*MAX_SPEED)
            while True:
            		bearing = getBearingInDegrees()
                if bearing > target_bearing:
            				self.move(0)  
                    break	
                    
				if target_bearing < initial_bearing:
        		self.left_motor.setVelocity(-0.1*MAX_SPEED)
            self.right_motor.setVelocity(0.1*MAX_SPEED)  
            while True:
            		bearing = getBearingInDegrees()
                if bearing < target_bearing:
            				self.move(0)  
                    break	                
        
            
            
    		
# create the Robot instance.
robot = Mybot()

robot.move_forward(.5)

i = 0
fin1 = 0

while robot.step(TIME_STEP) != -1: #@Cindy can you possibly do this? I don't follow what the code is doing
		Values = robot.getSensorValues()
		sensorValuesScan.push_back(Values) ####I DON'T KNOW WHAT THIS DOES OR WHAT PYTHON EQUIVALENT IS  
    
    if i ==0:
    		currentBearing = robot.get_bearing_in_degrees()
        robot.rotate_theta(355,currentBearing)
      
  	blockBearings = robot.get_block_bearings()
    rotate_until_bearing(blockBearing[0],robot.get_bearing_in_degrees)
